### PR TITLE
eagerly restore container execution state

### PR DESCRIPTION
**Pull request by danielnorberg**
Before starting the scheduler, poll for execution container state in order to record the completion of any containers that succeeded while styx was offline.

This avoids styx unnecessarily failing successfully completed executions due to having artifically "timed out" when starting up after downtime.

### CODE REVIEW

**Comment by danielnorberg** [comment]
Would be good to get in some tests of this new behavior in this PR but I'm opening it now to validate the general approach.

**Comment by danielnorberg** [comment]
@fabriziodemaria PTAL

File: pom.xml [hunk 1]
```diff
@@ -86,6 +86,11 @@
         <version>19.0</version>
       </dependency>
       <dependency>
+        <groupId>com.github.rholder</groupId>
+        <artifactId>guava-retrying</artifactId>
```
Comment (honnix) [inline pom.xml]: Good to know this lib. We can replace own simple retry code (storage related) with this.

File: styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java [hunk 2]
```diff
@@ -210,6 +216,21 @@ public void close() throws IOException {
...
+    final Retryer<Object> retryer = RetryerBuilder.newBuilder()
+        .retryIfException()
+        .withWaitStrategy(WaitStrategies.exponentialWait())
+        .withStopStrategy(StopStrategies.stopAfterAttempt(3))
+        .build();
+
+    try {
+      retryer.call(Executors.callable(this::tryPollPods));
+    } catch (ExecutionException | RetryException e) {
+      throw new RuntimeException(e);
```
Comment (honnix) [inline styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java]: This is stop Styx starting process. Is it intended?

File: styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java [hunk 3]
```diff
@@ -246,20 +267,24 @@ void examineRunningWFISandAssociatedPods(PodList podList) {
...
-        logEvent(Watcher.Action.MODIFIED, pod, resourceVersion, true);
-        inspectPod(Watcher.Action.MODIFIED, pod);
-      }
+      tryPollPods();
     } catch (Throwable t) {
       LOG.warn("Error while polling pods", t);
     }
   }
 
+  private synchronized void tryPollPods() {
```
Comment (honnix) [inline styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java]: Does it have to be `synchronized`? `DockerRunner` is created in a very lazy way.

File: styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java (hunk 2 above)
Comment (danielnorberg) [inline styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java]: Yes, if styx is unable to contact k8s at startup I believe it should crash and be restarted.

File: styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java (hunk 3 above)
Comment (danielnorberg) [inline styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java]: Yes, it needs to either be `synchronized` or `restore` needs to enqueue a call to `tryPollPods` on the `EXECUTOR` and introduce a subtle (and easily violated) requirement that the executor is single threaded. I don't believe `inspectPod` and `tryPollPods` are thread safe.

Comment (danielnorberg) [inline styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java]: As for the lazy creation of `DockerRunner`, this PR changes that behavior to eagerly instantiate the docker runner in `StyxScheduler.restoreState`: https://github.com/spotify/styx/pull/137/files#diff-a3fabd4f5e1e00cf48f40dfdfa3356f1R405

The lazy docker runner instantiation mechanism itself is a bit confusing and should perhaps be changed. Perhaps that could be a separate PR though.

Comment (honnix) [inline styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java]: Yeah, you are right. Disregard my comment. It feels like some design thoughts behind `RoutingDockerRunner`, not sure what it exactly is. There could be some reason making it lazy.

**Comment by honnix** [comment]
Overall LGTM.

File: styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java (hunk 2 above)
Comment (fabriziodemaria) [inline styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java]: The connection towards k8s should already be setup correctly at this point, and `restore` could be treated as a best effort procedure instead of crashing. I don't have a strong opinion on this, both approaches are fine.

**Comment by fabriziodemaria** [comment]
@danielnorberg as you mentioned in a comment above, it would be good to have a test around this, but overall looks good.

File: styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java (hunk 2 above)
Comment (danielnorberg) [inline styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java]: The connection might well be setup/configured correctly but connections and requests can and do fail at any time. I would rather crash and retry instead of a best effort `restore` that might lose those completed executions. Also, we do not do best effort reads from bigtable and datastore when restoring state. It seems odd to treat k8s differently.

File: styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java (hunk 3 above)
Comment (danielnorberg) [inline styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java]: I guess the reason for the laziness has to do with the ability to reconfigure the target GKE cluster at runtime.

Comment (honnix) [inline styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java]: Ah OK, that `globalDockerRunnerId` thing. It doesn't have to be lazy.

**Comment by honnix** [comment]
+1

File: styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java (hunk 2 above)
Comment (bergman) [inline styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java]: Why is the retry logic needed at all? It's part of a startup routine, I think it's ok to fail hard here.

Is it expected to fail intermittently?

Comment (bergman) [inline styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java]: Hmm, how would completed executions be lost if we crash hard on first error?

Comment (danielnorberg) [inline styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java]: Crashing hard and restarting on first error implies replaying all bigtable events etc again, which is relatively expensive. Thus I chose to have retries here to cheaply tolerate temporary netsplits.
